design-proposals/2023-11-08-complex-animatable-interactive-annotations.md

```diff
@@ -0,0 +1,55 @@
+# Complex, Animatable, Interactive Annotations Proposal
+
+An annotation as described in this proposal is a piece of UI that is drawn on the map at a single coordinate. In Apple MapKit this would be an MKAnnotationView, and in Google world this would be similar to a Marker. Annotations are expected to be user interactable, animatable and allow for complex ui to be represented.
+
+## Motivation
+
+Most mobile map experiences display some sort of custom content on the map, often in the form of annotations. MapLibre’s current annotation system is difficult to use, performs poorly, and lacks many of the features developers expect from modern mapping libraries.
+
+This proposal introduces the concept annotations that are complex, animatable and interactive to maplibre’s core. The first phase allows bitmap backed annotations to be rendered. The second phase will enable native platform views (ios, android) to be rendered directly by maplibre-gl.
+
+These new annotations will provide a more flexible tool for drawing user content on the map that is more inline with what developers expect from the built-in mobile map toolkits. The focus on animations and interactivity allows developers to create differentiated map experiences that feel made for mobile.
+
+By introducing a specialized rendering path for annotations, we bypass some of the limitations of the current layer-based system. This leads to better performance, especially in scenarios with many dynamic elements, where full layout re-calculations would otherwise cause noticeable lag and degrade the user experience.
+
+The customization options that these annotations will provide, together with the existing MapLibre stack, aim to offer a system that is more adaptable and robust than those from Google or Apple.
+
+## Proposed Change
+
+### Phase 1
+For this phase we would like to just build our support for bitmaps. The idea being that all native views eventually become bitmaps and supporting them robustly will eventually allow the renderer to draw native views correctly.
+
+* Introduce “Annotation” as a first class citizen of maplibre-gl core c++ layer. This feature would only be available to native clients and platforms and not maplibre-js.
+* Annotations rely on their own bitmap handling and not on the symbol layer texture atlas. Bitmaps should be able to be transformed and quickly swappable to support animations.
+* Annotations can define a label (reuse symbol layer logic)
```

Comment:
```suggestion
* Annotations can define a label (reuse symbol layer logic).
```

What does this mean exactly? Do you mean that Annotations can *have* a label?

Reply:
Right, and this is one of tricky parts. It makes sense to separate out the rendering of the label and the rendering of the annotation content. Labels need to know about the rest of the map to best position themselves (or in some cases hide themselves). Most of the logic exists in some form already inside of SymbolLayer, so it should be possible to optionally allow Annotations to opt into displaying a label.